src/boot_O2/__osMalloc.c

```diff
@@ -45,181 +49,353 @@ ArenaNode* ArenaImpl_GetLastBlock(Arena* arena) {
+ *
+ * @param arena  The Arena to check.
+ * @return u8    `true` if the \p arena has been initialized. `false` otherwise.
+ */
 u8 __osMallocIsInitalized(Arena* arena) {
     return arena->isInit;
 }
 
+/**
+ * @brief Allocates at least \p size bytes of memory using the given \p arena.
+ * The block of memory will be allocated from the beginning of \p arena.
+ *
+ *  - If there's not enough space in the given \p arena, this function will fail, returning `NULL`.
+ *  - If \p size is zero, then an empty region of memory is returned.
+ *
+ * To avoid memory leaks, the returned pointer should be eventually deallocated using either `__osFree` or
+ * `__osRealloc`.
+ *
+ * @param[in, out] arena  The specific Arena to be used for the allocation.
+ * @param[in] size        The size in bytes that will be allocated.
+ * @return void*          On success, the allocated area of the \p arena memory. Otherwise, `NULL`.
+ */
 void* __osMalloc(Arena* arena, size_t size) {
     ArenaNode* iter;
     ArenaNode* newNode;
-    void* alloc;
-    u32 blockSize;
-    alloc = NULL;
+    void* alloc = NULL;
 
     size = ALIGN16(size);
+
     ArenaImpl_Lock(arena);
+
+    // Start iterating from the head of the arena.
     iter = arena->head;
 
+    // Iterate over the arena looking for a big enough space of memory.
     while (iter != NULL) {
         if (iter->isFree && iter->size >= size) {
-            ArenaNode* next;
-            blockSize = ALIGN16(size) + sizeof(ArenaNode);
+            size_t blockSize = ALIGN16(size) + sizeof(ArenaNode);
+
+            // If the block is larger than the requested size, then split it and just use the required size of the
+            // current block.
             if (blockSize < iter->size) {
-                newNode = (ArenaNode*)((u32)iter + blockSize);
+                ArenaNode* next;
+
+                newNode = (ArenaNode*)((uintptr_t)iter + blockSize);
                 newNode->next = iter->next;
                 newNode->prev = iter;
                 newNode->size = iter->size - blockSize;
-                newNode->isFree = 1;
+                newNode->isFree = true;
                 newNode->magic = NODE_MAGIC;
 
                 iter->next = newNode;
                 iter->size = size;
+
                 next = newNode->next;
-                if (next) {
+                if (next != NULL) {
                     next->prev = newNode;
                 }
             }
 
-            iter->isFree = 0;
-            alloc = (void*)((u32)iter + sizeof(ArenaNode));
+            iter->isFree = false;
+            alloc = (void*)((uintptr_t)iter + sizeof(ArenaNode));
             break;
         }
 
         iter = iter->next;
     }
+
     ArenaImpl_Unlock(arena);
 
     return alloc;
 }
 
+/**
+ * @brief Allocates at least \p size bytes of memory using the given \p arena.
+ * Unlike __osMalloc, the block of memory will be allocated at the end of the \p arena.
```

Comment:
```suggestion
 * Unlike __osMalloc, the block of memory will be allocated from the end of the \p arena.
```
In a similar vein to above, "at" suggests to me at first glance that it would be placed at the very end of the arena, rather than the last sufficiently large free block.